### PR TITLE
MAINT: Return size_t from num_codepoints in string ufuncs Buffer class

### LINKED ISSUE #25565: MAINT: Return size_t from num_codepoints in Buffer class in string ufuncs
So that we don't have to worry about length getting negative or underflowing.

Proposed in https://github.com/numpy/numpy/pull/25558#discussion_r1447654922.

### PR DESCRIPTION
**Pull request by lysnikolaou**
Closes #25565.

### CODE REVIEW

**Comment by ngoldbaum** [comment]
Looks like there's a heap buffer overflow in `rstrip`. Also woohoo that's the first time I've seen the compiler sanitizer job catch a bug like that.

**Comment by mhvk** [comment]
The sanitizer actually did a really nice job on my fft via gufunc PR (#25536). Though it still took me quite a while to figure out where I was writing a byte too many! But very nice that it found it, since everything else just passed fine.

File: numpy/_core/src/umath/string_ufuncs.cpp [hunk 1]
```diff
@@ -150,11 +150,11 @@ template <ENCODING enc>
...
 string_find(Buffer<enc> buf1, Buffer<enc> buf2, npy_int64 start, npy_int64 end)
 {
-    npy_int64 len1 = buf1.num_codepoints();
-    npy_int64 len2 = buf2.num_codepoints();
+    size_t len1 = buf1.num_codepoints();
+    size_t len2 = buf2.num_codepoints();
 
     adjust_offsets(&start, &end, len1);
-    if (end - start < len2) {
+    if (end - start < static_cast<npy_int64>(len2)) {
```
[the remark this answers is not in the document]
Comment (lysnikolaou) [reply numpy/_core/src/umath/string_ufuncs.cpp:157]: My understanding is that the signedness of `start + len2` is implementation-specific and depends on the size of `size_t`. For implementations where it's a 64-bit integer, the resulting type is also unsigned, so we end up with signed-unsigned comparison again.

> If the unsigned type has conversion rank greater than or equal to the rank of the signed type, then the operand with the signed type is implicitly converted to the unsigned type.
> 1) the ranks of all signed integer types are different and increase with their precision: rank of signed char < rank of short < rank of int < rank of long int < rank of long long int
> 2) the ranks of all signed integer types equal the ranks of the corresponding unsigned integer types

**Comment by lysnikolaou** [comment]
The test failures are unrelated as far as I can tell.

File: numpy/_core/src/umath/string_ufuncs.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (mhvk) [reply numpy/_core/src/umath/string_ufuncs.cpp:157]: Well, maybe explicit is better, then...

**Comment by mhvk** [comment]
Rerunning the pypy job just in case. Is it failing also for other PRs?

**Comment by lysnikolaou** [comment]
I don't see any other failures, no, but I think it was a very early segfault in pypy code before the build even began. Could you maybe rerun the `OpenSUSE Netlib BLAS/LAPACK` one?

**Comment by mhvk** [comment]
Ah, sorry, missed that there was another failure. Let's see what a rerun does.

**Comment by lysnikolaou** [comment]
Yup, they're both okay now.

**Comment by ngoldbaum** [comment]
There seems to be a new 60 minute build time limit we're hitting in azure, it's unrelated to this PR.

**Comment by ngoldbaum** [comment]
Thanks for this!

**Pull request merged**